### PR TITLE
Adding min dom set

### PR DESCRIPTION
**Pull request by chadbrewbaker**
Requires an upgrade to a more recent Z3 that has the Optimize package.

Thanks for the graph coloring example!

### CODE REVIEW

File: min_dom_set_z3.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        build_peternson_3_coloring_graph(),
+        build_fat_graph()
+    ]
+
+    for G, name, layout in Gs:
+        print 'Trying to min dom set  %s now (%d nodes, %d edges)..' % (repr(name), G.number_of_nodes(), G.number_of_edges())
+        t1 = time.time()
+        s = min_dom_set(G)
+        t2 = time.time()
+        print 'OK, found a solution with %d dominators' % sum( s.values())
```
Comment (0vercl0k) [inline min_dom_set_z3.py]: Extra white-space

File: min_dom_set_z3.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    """Try to dominate the graph with the least number of verticies possible"""
+    s = Optimize()
+    nodes_colors = dict((node_name, Int('k%r' % node_name)) for node_name in graph.nodes())
+    for node in graph.nodes():
+           s.add(And(nodes_colors[node] >= 0, nodes_colors[node] <= 1))
+           dom_neighbor  = Sum ([ (nodes_colors[j]) for j in graph.neighbors(node) ])
+           s.add(Sum(nodes_colors[node], dom_neighbor ) >=1)
+    s.minimize( Sum([ nodes_colors[y] for y in graph.nodes()]) )
+
+    if s.check() == unsat:
```
Comment (0vercl0k) [inline min_dom_set_z3.py]: Could you == sat instead as the pass statement if a bit useless

File: min_dom_set_z3.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+def min_dom_set(graph):
+    """Try to dominate the graph with the least number of verticies possible"""
+    s = Optimize()
+    nodes_colors = dict((node_name, Int('k%r' % node_name)) for node_name in graph.nodes())
+    for node in graph.nodes():
+           s.add(And(nodes_colors[node] >= 0, nodes_colors[node] <= 1))
+           dom_neighbor  = Sum ([ (nodes_colors[j]) for j in graph.neighbors(node) ])
+           s.add(Sum(nodes_colors[node], dom_neighbor ) >=1)
+    s.minimize( Sum([ nodes_colors[y] for y in graph.nodes()]) )
```
Comment (0vercl0k) [inline min_dom_set_z3.py]: Sounds like there is a missing white-space after graph.nodes()

File: min_dom_set_z3.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+# -*- coding: utf-8 -*-
+
+import time
+import pygraphviz as pgv
+from z3 import *
+import sys
+
+def min_dom_set(graph):
+    """Try to dominate the graph with the least number of verticies possible"""
+    s = Optimize()
```
Comment (0vercl0k) [inline min_dom_set_z3.py]: Awesome! I didn't even know this feature - maybe I'll see if it's worth making a simple example for the essentials directory :+1:

File: min_dom_set_z3.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    nodes_colors = dict((node_name, Int('k%r' % node_name)) for node_name in graph.nodes())
+    for node in graph.nodes():
+           s.add(And(nodes_colors[node] >= 0, nodes_colors[node] <= 1))
+           dom_neighbor  = Sum ([ (nodes_colors[j]) for j in graph.neighbors(node) ])
+           s.add(Sum(nodes_colors[node], dom_neighbor ) >=1)
```
Comment (0vercl0k) [inline min_dom_set_z3.py]: Extra white-space after dom_neighbor, and it's missing one after >=

File: min_dom_set_z3.py (hunk 4 above)
Comment (0vercl0k) [inline min_dom_set_z3.py]: What do you think about adding a link to http://rise4fun.com/Z3Opt/tutorial/guide? Also worth mentioning this is available from >=4.4.1 only?

File: min_dom_set_z3.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
```
Comment (0vercl0k) [inline min_dom_set_z3.py]: Wanna add your name / twitter in case somebody wants to ping you about it?

**Comment by 0vercl0k** [comment]
Actually; I was thinking about this yesterday night - what do you think about integrating the min_dom_set function directly inside the graph_color example?

It would be a pretty cool way to see what Optimize gives us, instead of trying the number of color step-by-step. What do you thing @chadbrewbaker ?

Cheers

**Comment by chadbrewbaker** [comment]
Ok I will merge the two scripts into one,  and throw in a twitter comment blurb on the min_dom_set function.

I'm impressed that SMT solvers have gotten so good. A decade ago you had to write branch and bound codes in C for these things.

**Comment by chadbrewbaker** [comment]
Merged coloring and min-dom set. Unfortunately Z3 doesn't have a Max() operator over a lists like it does for Sum(), so encoding graph coloring as an optimization problem is kind of hard.

**Comment by 0vercl0k** [comment]
Perfect @chadbrewbaker ; I've polished a bit the script ([https://github.com/0vercl0k/z3-playground/commit/fa991a3acedf49ac5309f54787b96366df4709c6](https://github.com/0vercl0k/z3-playground/commit/fa991a3acedf49ac5309f54787b96366df4709c6)) & tried that out on [https://github.com/0vercl0k/z3-playground/commit/7cf9bd87214eb81b4ee640ca70c245a66aed8287](https://github.com/0vercl0k/z3-playground/commit/7cf9bd87214eb81b4ee640ca70c245a66aed8287) - really cool :-).

**Pull request merged**